mvpa2/measures/label.py

```diff
@@ -0,0 +1,139 @@
+# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the PyMVPA package for the
+#   copyright and license terms.
+#
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Label adjacent features in a dataset
+"""
+import numpy as np
+from collections import defaultdict
+from operator import itemgetter
+
+from ..base.param import Parameter
+from ..base.constraints import EnsureBool
+from ..datasets.base import Dataset
+from ..misc.neighborhood import CachedQueryEngine
+
+from .base import Measure
+
+if __debug__:
+    from ..base import debug
+
+
+class ClustersLabeler(Measure):
+    """Determines contiguous regions of features using a query engine.
+
+    "label" in the name chosen to mimic scipy.ndimage.measurements.label
+    """
+    qe = Parameter(
+        None,
+        doc="QueryEngine to determine neighbors of any given feature")
```

Comment:
1. queryengine here is pretty much merely a generalization to define immediate neighborhood.  E.g. before `scipy.ndimage.measurements.label` internally was defining by default "immediate neighbors touching the face" and was restricted only to volume.  Here use of query engine relaxes it and allows to find an arbitrary immediate (or not if e.g. Sphere with big radius is used) neighborhood (in volume or on surface).  But in general/default you would just use `Sphere(1)` making it identical to how previous implementation behaved (using `scipy...label`)
2. interesting point on per-subject QEs!  Similar aspect was recently dealt with in searchlight hyperalignment (slhyper below) implementation (see e.g. https://github.com/PyMVPA/PyMVPA/blob/master/mvpa2/algorithms/searchlight_hyperalignment.py#L232 and where used in there).  But situation there was "easier" and somewhat clearer due to that slhyper obtaining a list of datasets, so they could easily carry varying .fa as per each subject and were not required to have corresponding features... Here we have inherent assumption that features already correspond across subjects, since we generally just computing a mean performance map, and clustering/qe is used later only to figure out clusters on that map.  So unless we want to aim for some extended functionality where QEs would also be used to project/mean performances within neighborhoods of each subject, there is probably no need for per-subject QE, but even then such meaning could simply be done prior, and for this computation clustering is done on e.g. a corresponding "template" surface